torch/onnx/_internal/exporter/_torchlib/ops/hop.py

```diff
@@ -0,0 +1,90 @@
+"""Implementation for higher-order operators."""
+
+from __future__ import annotations
+
+from typing import Sequence
+
+import torch
+from torch.onnx._internal._lazy_import import onnxscript_ir as ir
+from torch.onnx._internal.exporter import _core
+from torch.onnx._internal.exporter._torchlib._torchlib_registry import onnx_impl
+
+
+def call_op(
+    op_type: str,
+    *args: ir.Value,
+    _num_outputs: int = 1,
+    _domain: str = "",
+    **kwargs: int | float | str | bool | ir.Graph | ir.TensorProtocol,
+) -> Sequence[ir.Value]:
+    """Call an operator with the given arguments and keyword arguments.
+
+    Arguments are always inputs, while keyword arguments are attributes.
+    """
+    # This is a wrapper around the IR node creation that hooks into the _builder.OpRecorder
+    # tracer so that all nodes created are recorded the same way as if we were to use
+    # onnxscript ops directly.
+    from onnxscript.ir import convenience as ir_convenience
+
+    assert _core.current_tracer is not None
+    tracer = _core.current_tracer
+
+    inputs = list(args)
+
+    # If final inputs are None, strip them from the node inputs
+    for input in reversed(inputs):
+        if input is not None:
+            break
+        inputs.pop()
+
+    # Construct and filter out None attributes
+    attributes = [
+        attr
+        for attr in ir_convenience.convert_attributes(kwargs)
+        if attr.value is not None  # type: ignore[union-attr]
+    ]
+    tracer.nodes.append(
+        node := ir.Node(
+            _domain,
+            op_type,
+            inputs=inputs,
+            attributes=attributes,
+            num_outputs=_num_outputs,
+            version=tracer.opset.version,
+        )
+    )
+    return node.outputs
+
+
+@onnx_impl(torch.ops.higher_order.cond)
+def higher_order_cond(
+    cond: ir.Value,
+    true_func: ir.Function,
+    false_func: ir.Function,
+    inputs: Sequence[ir.Value],
+) -> Sequence[ir.Value]:
+    then_node = ir.Node(
+        true_func.domain, true_func.name, inputs, num_outputs=len(true_func.outputs)
+    )
+    else_node = ir.Node(
+        false_func.domain, false_func.name, inputs, num_outputs=len(false_func.outputs)
+    )
+
+    # ONNX Runtime complains about duplicate output names if we don't rename them.
+    # But the doesn't seem to be an actual violation of SSA form without renaming.
+    for out in then_node.outputs:
+        out.name = f"{out.name}_{true_func.name}"
+    for out in else_node.outputs:
+        out.name = f"{out.name}_{false_func.name}"
+
+    return call_op(
+        "If",
+        cond,
+        _num_outputs=len(true_func.outputs),
+        then_branch=ir.Graph(
```

Comment:
I'm wondering what happens to the model weights used by these functions in case the model is bigger than 2Gb but that may be the topic of another PR.

Reply:
the model weights are always passed in as inputs so we don't actually store any weights in the subgraph or the functions.